### ISSUE TITLE
Node Binding Portability

### ISSUE THREAD

**Issue by allierowan** [valhalla/valhalla#1418]
## What does it mean for node bindings to be portable?

When we build node bindings, we generate a binary `.node` file with the executable for the bindings. As with any binary, the `.node` binary has dependencies (for example, boost, protobuf, etc.). For all the functions that we use that come from dependencies, we need to either tell the binary to go looking for the third party code on the system dynamically, or we can build all of the functions and their dependent code into the binary. A `statically linked` dependency is included in the binary. A `dynamically linked` dependency needs to live on the environment where the binary is run. The "portability" of a binary is judged by whether or not you can run it on a system that doesn't already have the dependencies locally. 

## Different levels of portability

Even if a package doesn’t include every single dependency when creating the binary, it could still function in a portable manner, depending on the environment it is being run on. Some systems may include some of the basic dependencies by default, and all systems are going to function differently. 

It would be great to get the node bindings for Valhalla to be extremely portable, to the point where they could be run on a totally different Linux distribution like Amazon linux so that Valhalla could be run on Lambda or Spark.

Here are the different types of portability we can work toward:

- get binary working on the same linux distro with the same dependencies installed. This is done! :tada:
- get the binary working on the same linux distro with non of the dependencies actually installed (so a vanilla system)
-   get the binary working on OSX with none of the dependencies installed
-   get the binary working on a totally different Linux distribution like Amazon linux so that valhalla can be run on Lambda or Spark 

## Implications/Priorities

I think that it goes without saying that we _eventually_ want to get to total portability, but this is pretty time-consuming work that involves a lot of staring at cryptic errors and figuring out how to include different dependencies statically. The question I have is, what is our priority for portability? Right now, we have successfully seen the node bindings work if run in a docker image that matches the linux image we use on circle. This, in theory, is enough for local development, but is maybe a little bit clumsy. 

So the main question I have is: do we want to get the binaries portable enough that we can develop locally on OSX outside of docker and on a clean linux distro, or are we okay working from a docker image for awhile? What level of portability do we want to shoot for in the near term? Please share your thoughts on this ticket!

cc @mapbox/directions-v6 and also @springmeyer who helped me think about this!

**Comment by karenzshea** [comment]
cc @danpat

**Comment by kevinkreiser** [comment]
if one of the main priorities here is local development, i would recommend building the valhalla node module locally (its not portable) and then developing your node app against that.

this is how i work on python applications that use valhalla. there is no need to use docker etc but there is just a little bit of specifying paths so that python can see my locally built valhalla python module. does something like this exist to tell node to look in certain directories first when loading modules so that it catches your local valhalla module? it must right because of the node tests in valhalla!

the main reason i think this is preferable for local development is because it allows you to both change your node/python program and your valhalla node/python module when you find stuff that needs to be changed. nothing is more frustrating than waiting to continue development on your node/python app because you first have to publish a node/python module.

basically if you do get full portability you still have a very slow loop if you find a problem in the module itself. for a tighter development loop i recommend developing ALL THE THINGS locally first. then when you are done with your development publish ALL THE THINGS :smile: 

just my 2 cents!

**Comment by allierowan** [comment]
We have made a good step toward this goal in closing #1423. This PR made the bindings portable on mac. I think down the line, we will want to work toward a more "gold-standard" portability, so I will leave this open for now.

**Comment by aindeev** [comment]
This would be a wonderful addition. We just attempted to use Valhalla node bindings on a linux machine and ran into these exact problems. We are attempting to build and publish a statically linked version to a private repository now.

Would making the node bindings statically linked be as simple as adding:
```
-DBUILD_SHARED_LIBS=Off -DENABLE_PYTHON_BINDINGS=Off -DENABLE_DATA_TOOLS=OFF -DENABLE_SERVICES=OFF -DBoost_USE_STATIC_LIBS=ON -DProtobuf_USE_STATIC_LIBS=ON -DLZ4_USE_STATIC_LIBS=ON
```
to the `build-release-binary-linux` step?
